feat(buzzer): Add `esp-hal-buzzer` to drive a piezo-electric buzzer.

AnthonyGrondin opened the pull request:
## Thank you for your contribution!

We appreciate the time and effort you've put into this pull request.
To help us review it efficiently, please ensure you've gone through the following checklist:

### Submission Checklist 📝
- [x] I have updated existing examples or added new ones (if applicable).
- [x] I have used `cargo xtask fmt-packages` command to ensure that all changed code is formatted correctly.
- [x] My changes were added to the [`CHANGELOG.md`](https://github.com/esp-rs/esp-hal/blob/main/esp-hal/CHANGELOG.md) in the **_proper_** section.
- [x] My changes are in accordance to the [esp-rs API guidelines](https://github.com/esp-rs/esp-hal/blob/main/documentation/API-GUIDELINES.md)

#### Extra:
- [x] I have read the [CONTRIBUTING.md guide](https://github.com/esp-rs/esp-hal/blob/main/documentation/CONTRIBUTING.md) and followed its instructions.

### Pull Request Details 📖

#### Description
This is a resurrection for #814 that was closed by mistake a while ago. The intent is to provides a driver for a piezo-electric buzzer by abstracting LEDC and offering a user-friendly API to control the buzzer 

Since the GPIO refactoring, I've solved the issue I had with providing a type for an optional volume pin.

#### Testing
An example was added in [examples/src/bin/buzzer.rs](https://github.com/AnthonyGrondin/esp-hal/blob/8391db1d6e2226f340daa3358a3a47e54194c9d1/examples/src/bin/buzzer.rs) and works on my hardware. The duty volume is working on the specific hardware that I have, using logic gates to decrease the output.

bugadani commented on esp-hal/src/ledc/timer.rs:
```diff
@@ -91,6 +91,43 @@ pub mod config {
+                #[cfg(esp32)]
+                20 => Self::Duty20Bit,
+                #[cfg(not(esp32))]
+                _ => panic!("Invalid u32 value for Duty. {} not in range 1-14", value),
```
If the conversion can fail, it should probably be a `TryFrom` impl.

bugadani commented on examples/Cargo.toml:
```diff
@@ -32,6 +32,7 @@ embedded-storage    = "0.3.1"
 esp-alloc           = { path = "../esp-alloc" }
 esp-backtrace       = { path = "../esp-backtrace", features = ["exception-handler", "panic-handler", "println"] }
 esp-hal             = { path = "../esp-hal", features = ["log"] }
+esp-hal-buzzer      = { path = "../esp-hal-buzzer", optional = true }
```
I don't think I find this integral enough to be included in `examples`. Also I think xtasks is prepared to run examples that are placed elsewhere, e.g. in the `esp-hal-buzzer` crate.

AnthonyGrondin replied:
I added it there because `esp-hal-smartled` also has an example there. That being said, I can move the extra crates into a special `extras/` to not complicate the root with helper crates that are not part of the core HAL.

bjoernQ commented on examples/src/bin/buzzer.rs:
```diff
@@ -0,0 +1,61 @@
+        &clocks,
+    );
+
+    buzzer.play_song(songs::DOOM).unwrap();
+    buzzer.play_song(songs::FURELISE).unwrap();
+    buzzer.play_song(songs::MERRY_CHRISTMAS).unwrap();
+    buzzer.play_song(songs::MII_CHANNEL).unwrap();
+    buzzer.play_song(songs::NEVER_GONNA_GIVE_YOU_UP).unwrap();
+    buzzer.play_song(songs::ODE_TO_JOY).unwrap();
+    buzzer.play_song(songs::PACMAN).unwrap();
+    buzzer.play_song(songs::STAR_WARS).unwrap();
+    buzzer.play_song(songs::SUPER_MARIO_BROS).unwrap();
+    buzzer.play_song(songs::TAKE_ON_ME).unwrap();
+    buzzer.play_song(songs::TETRIS).unwrap();
+    buzzer.play_song(songs::THE_LION_SLEEPS_TONIGHT).unwrap();
+    buzzer.play_song(songs::ZELDA_LULLABY).unwrap();
+    buzzer.play_song(songs::ZELDA_THEME).unwrap();
```
I'm not quite sure if these songs should be contained in the library - the macro to define them of course but maybe the songs themselves would be better to live in the examples?

jessebraham commented:
Sorry for hijacking your PR for this, but I guess we need to have some discussion with regards to where packages like this belong. We have `esp-hal-smartled` in the repo, but frankly it's just a bit of an annoyance from the perspective of maintenance/publishing, and barely anybody seems to even use it based off the [download stats on crates.io](https://crates.io/crates/esp-hal-smartled). I'm not sure we really want to take on a number of packages like this given their limited use, though I do acknowledge their utility.

AnthonyGrondin commented:
I agree with this. I see it like a chicken :chicken: and egg :egg: situation. Such extra packages are an unnecessary burden because nobody uses them, but nobody uses them because such package doesn't exist.

My original intent was to add the Buzzer as a new driver in the HAL. But since it's not technically a hardware driver, but rather a combination of drivers, it was suggested to move the buzzer driver in its own specific package.

I agree with moving both `esp-hal-buzzer` and `esp-hal-smartled` into `extras/`, and I think we should treat those packages as tier 2 packages, having a lower priority, maintenance, etc. Tier 1 packages being the most important ones, part of the root of the repo.

jessebraham commented:
Sorry for the delay here, it took some time for us to discuss and decide on how we wanted to proceed.

We have decided to create the [esp-hal-community](https://github.com/jessebraham/esp-hal-community/) repository, to which I have already moved `esp-hal-smartled` and its associated example. If you wouldn't mind opening a new PR there adding this package, that would be much appreciated!

I have not really documented the contribution process there yet (though I did open [an issue](https://github.com/esp-rs/esp-hal-community/issues/2)), but I hope it is intuitive enough. Feel free to ping me if you need any assistance :)

AnthonyGrondin commented:
- I've moved `esp-hal-buzzer` to `esp-hal-community` in https://github.com/esp-rs/esp-hal-community/pull/3
- I've updated this PR title and description to reflect the changes
- I rebased on the latest HEAD

bjoernQ reviewed (approved):
LGTM
